bazel/external/requirements.in

```diff
@@ -1,43 +1,44 @@
 # requirements.in
 setuptools==49.6.0
-grpcio-tools
-grpcio
-redis>=3.5
-protobuf
-six
-pyyaml
-urllib3
-certifi
-sentry-sdk
+grpcio-tools>=1.46.3
+grpcio>=1.47.0
+redis==3.5.3
+protobuf==3.20.1
+six>=1.16.0
+pyyaml==5.3.1
+urllib3==1.26.11
+certifi>=2022.6.15
+sentry-sdk>=1.9.0
 snowflake>=0.0.3
 #  prometheus_client==0.3.1 required (see e.g. magmad/tests/metrics_tests.py)
 prometheus_client==0.3.1
-redis-collections
-python-redis-lock
-jsonpickle
-netifaces
-scapy
-wsgiserver
-flask
-bravado_core
+redis-collections==0.11.0
+python-redis-lock>=3.7.0
+jsonpickle>=2.2.0
+netifaces>=0.11.0
+scapy==2.4.5
+wsgiserver>=1.3
+flask==1.1.4
+markupsafe==1.1.1
+bravado_core==5.17.0
 #  jsonschema version 4 not compatible with magma swagger specs
-jsonschema==3.1.0 
-psutil
-systemd
-cryptography
+jsonschema==3.2.0
+psutil==5.9.1
+systemd-python>=234
+cryptography==3.2.1
 #  h2>=3,<4 is requirement of aioh2 (loaded via bazel)
-h2>=3,<4
+h2==3.2.0
```

Comment:
What is the reason for pinning most of the versions in `requirements.in`? As far as I understand the requirements.in, it is supposed to contain version constraints, and then `pip-compile` generates the `requirements.txt` which has exact pinnings. The `requirements.txt` will then be used by `pip` to install the exactly pinned versions.

This line for example would previously tell `pip-compile` to not upgrade to a version `>= 4.0`. With the line changed to `h2==3.2.0`, `pip-compile` no longer knows about this constraint.

Without the exact pinning, I can use other package repositories like nixpkgs to install Python dependencies. This stops working when we exactly pin the dependencies, because that very specific combination of packages usually isn't available there.

Reply:
Related question: Why do we convert some of the `>=` lines to `==`, while others stay `>=` constraints?

Reply:
I some cases (e.g. jinja2), things break if you allow `>=` and pip-compile won't be able to resolve it. Not sure about h2 in particular.

Reply:
Ah, i see, thanks!

Reply:
Thanks, I am just copying below what I sent privately on slack:

"I had another look into the [requirements.in](http://requirements.in/) file and how we compile this to a requirements.txt. One thing I have noticed today is that the output of pip-compile depends on what is already in the requirements.txt. In fact, if you delete the requirements.txt on the current state of master and try to build it again, it will fail (see screenshot). To resolve the version conflict in the screenshot, I manually set flask<2.1.0.  pip-compile apparently only changes versions in the requirements.txt if it is forced to. If the existing versions fulfil the [requirements.in](http://requirements.in/), pip-compile will not upgrade versions.
With all of that in mind, I changed the PR so that we only use fixed versions where they were already in place. I incremented all of the fixed versions, added the constraint on flask, and copied the sentry-sdk constraint from the setup.py file. Additionally, I changed systemd to systemd-python. These are actually different packages with similar interfaces. systemd-python is what is actually required in magma.
Finally, I removed the requirements.txt , built it again from scratch, and added the changes to the PR."

Reply:
Yes, that seems to be in accordance with https://github.com/jazzband/pip-tools#example-usage-for-pip-compile:

Reply:
> What is the reason for pinning most of the versions in `requirements.in`? 

That originated from a conversation between @LKreutzer and me, who brought https://github.com/magma/magma/issues/11743 to my attention. The second bullet point, according to our interpretation, implied that one wants to make sure, that same versions are used independently of whether you rely on setup.py or requirements.in.

But just as an explanation, I guess its fine without that as well.

Reply:
Good point, ensuring that installed versions are the same when using `pip install` vs building & installing the `.deb` package is a good argument for pinning in `requirements.in`.

Reply:
But I think the current change where we keep the pinning for packages that were already pinned is fine for now.